src/SecureStore.HashicorpVault/HashicorpVaultClient.cs

```diff
@@ -0,0 +1,182 @@
+                    };
+                default:
+                    return null;
+            }
+        }
+
+        public async Task<string> SetCredentialAsync(string secretName, Credential credential)
+        {
+            var keyVaultClient = GetVaultClient();
+            switch (_context.SecretsEngine)
+            {
+                case SecretsEngine.KeyValueV1:
+                    await keyVaultClient.V1.Secrets.KeyValue.V1.WriteSecretAsync(secretName, credential);
+                    return secretName;
+                case SecretsEngine.KeyValueV2:
+                    await keyVaultClient.V1.Secrets.KeyValue.V2.WriteSecretAsync(secretName, credential);
+                    return secretName;
+                case SecretsEngine.ActiveDirectory:
+                    throw new NotSupportedException("Writing secrets is not supported for Engine ActiveDirectory");
+                case SecretsEngine.Cubbyhole:
+                    var secretToSave = new Dictionary<string, object>
+                    {
+                        ["Username"] = credential.Username,
+                        ["Password"] = credential.Password,
+                    };
+                    await keyVaultClient.V1.Secrets.Cubbyhole.WriteSecretAsync(secretName, secretToSave);
+                    return secretName;
+                default:
+                    return null;
+            }
+        }
+
+        public Task DeleteCredentialAsync(string secretName) => DeleteSecretAsync(secretName);
+
+        private IVaultClient GetVaultClient()
+        {
+            IAuthMethodInfo authMethod;
+
+            switch (_context.AuthenticationType)
+            {
+                case AuthenticationType.AppRole:
+                    authMethod = new AppRoleAuthMethodInfo(_context.RoleId, _context.SecretId);
+                    break;
+                case AuthenticationType.UsernamePassword:
+                    authMethod = new UserPassAuthMethodInfo(_context.Username, _context.Password);
+                    break;
+                case AuthenticationType.Ldap:
+                    authMethod = new LDAPAuthMethodInfo(_context.Username, _context.Password);
+                    break;
+                case AuthenticationType.ClientCertificate:
+                    authMethod = new CertAuthMethodInfo(new X509Certificate2(_context.Certificate, _context.CertificatePassword));
+                    break;
+                case AuthenticationType.Token:
+                    authMethod = new TokenAuthMethodInfo(_context.Token);
+                    break;
+                default:
+                    throw new InvalidOperationException("Unsupported authentication type");
+            }
+
+            var vaultClientSettings = new VaultClientSettings(_context.VaultUri.ToString(), authMethod);
+            vaultClientSettings.VaultServiceTimeout = TimeSpan.FromSeconds(30); // TODO configurable
```

Comment:
Should we make the timeout configurable? I guess it does not make sense to do so, as that should not be a concern for the Orchestrator admin configuring this. We should just pick a good default and be done.

Reply:
Decided not to. It will be fixed at 30s. In the future, we will pass on a cancellation token.